### PR TITLE
Introduce appropriate clients for all our api groups

### PR DESCRIPTION
**Pull request by soltysh**
@fabianofranz David told me it's up to you to pick the names for those methods. These are meant to be used inside our CLI, instead of manually crafting the client for the appropriate api groups. 
Once we agree on the naming we'll need to go through our code and update it so that it uses the new methods.

@mfojtik @deads2k fyi

### CODE REVIEW

File: pkg/cmd/util/clientcmd/factory_client_access.go [hunk 1]
```diff
@@ -51,7 +63,22 @@ type ring0Factory struct {
...
+	OpenShiftAuthorizationClient() (authorizationclient.Interface, error)
+	OpenShiftBuildClient() (buildclient.Interface, error)
+	OpenShiftImageClient() (imageclient.Interface, error)
+	OpenShiftNetworkClient() (networkclient.Interface, error)
+	OpenShiftOAuthClient() (oauthclient.Interface, error)
+	OpenShiftProjectClient() (projectclient.Interface, error)
+	OpenShiftQuotaClient() (quotaclient.Interface, error)
+	OpenShiftRouteClient() (routeclient.Interface, error)
+	OpenShiftSecurityClient() (securityclient.Interface, error)
+	OpenShiftTeamplteClient() (templateclient.Interface, error)
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory_client_access.go]: typo

File: pkg/cmd/util/clientcmd/factory_client_access.go [hunk 2]
```diff
@@ -51,7 +63,22 @@ type ring0Factory struct {
 type ClientAccessFactory interface {
 	kcmdutil.ClientAccessFactory
 
+	// TODO: this should be removed when we finally get rid of pkg/client
 	Clients() (*client.Client, kclientset.Interface, error)
+	// use below clients instead whenever needed:
+	OpenShiftAppsClient() (appsclient.Interface, error)
```
Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory_client_access.go]: nit: I would rather have this as `OpenshiftInternalAppsClient` to be consistent with https://github.com/mfojtik/origin/blob/master/pkg/cmd/server/origin/controller/interfaces.go#L60

File: pkg/cmd/util/clientcmd/factory_client_access.go [hunk 3]
```diff
@@ -143,6 +170,150 @@ func (f *ring0Factory) Clients() (*client.Client, kclientset.Interface, error) {
 	return openShiftClient, kubeClientSet, nil
 }
 
+func (f *ring0Factory) OpenShiftAppsClient() (appsclient.Interface, error) {
```
Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory_client_access.go]: missing godoc ;-)

File: pkg/cmd/util/clientcmd/factory_client_access.go [hunk 4]
```diff
@@ -51,7 +63,22 @@ type ring0Factory struct {
 type ClientAccessFactory interface {
 	kcmdutil.ClientAccessFactory
 
+	// TODO: this should be removed when we finally get rid of pkg/client
 	Clients() (*client.Client, kclientset.Interface, error)
+	// use below clients instead whenever needed:
+	OpenShiftAppsClient() (appsclient.Interface, error)
+	OpenShiftAuthorizationClient() (authorizationclient.Interface, error)
```
Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory_client_access.go]: @deads2k do we need this client or we want upstream?

File: pkg/cmd/util/clientcmd/factory_client_access.go [hunk 5]
```diff
@@ -51,7 +63,22 @@ type ring0Factory struct {
 type ClientAccessFactory interface {
 	kcmdutil.ClientAccessFactory
 
+	// TODO: this should be removed when we finally get rid of pkg/client
```
Comment (mfojtik) [inline pkg/cmd/util/clientcmd/factory_client_access.go]: we still need core kube client and probably some other clients as well (deployments?) i would actually suggest, while you are here, to remove remove this function and create:

```
KubeInternalClient(name string) (kclientsetinternal.Interface, error)
DeprecatedOpenshiftClient(name string) (osclient.Interface, error)
```

Then fix the CLI so at least we can grep for the usage of Deprecated client and replace step-by-step :)

File: pkg/cmd/util/clientcmd/factory_client_access.go (hunk 4 above)
Comment (deads2k) [reply pkg/cmd/util/clientcmd/factory_client_access.go]: > @deads2k do we need this client or we want upstream?

This client points to our resources and not all of them will move upstream.

File: pkg/cmd/util/clientcmd/factory_client_access.go (hunk 5 above)
Comment (deads2k) [reply pkg/cmd/util/clientcmd/factory_client_access.go]: > we still need core kube client and probably some other clients as well (deployments?) i would actually suggest, while you are here, to remove remove this function and create:

The kube client shoudl already be accessible on the stock factory.

**Comment by fabianofranz** [comment]
Do we really need the `OpenShift` prefix? Not a strong preference and I'm ok if people think that makes it much clearer, but in most places we kind of been doing the opposite (prefix with `Kube` or `kube` or `k` **if** disambiguation is needed).

**Comment by deads2k** [comment]
> Do we really need the OpenShift prefix?

I think so.  We conflict on apps, authentication, and authorization.

**Comment by deads2k** [comment]
> KubeInternalClient(name string) (kclientsetinternal.Interface, error)

This one should already exist upstream



> DeprecatedOpenshiftClient(name string) (osclient.Interface, error)

This is a good idea if doesn't ripple too far.

**Comment by soltysh** [comment]
This is being picked in #16451.

**Pull request closed without merging**